ls1intum/Artemis
# Handle time-outs and other errors when submitting exams

### Checklist

### Motivation and Context

To the user it is currently not visible that Artemis is doing work in the background. Additionally we need to make sure it is possible to submit again when a time-out or different error occurs.

### Description

Introduces a client-side time-out (20s) and re-enables submission if that or an error occurs. Adds support for properly displaying 409 errors and introduces a spinner for when the REST call in the background loads. The "Continue" button is also disabled while loading.

### Steps for Testing

Create an Exam and test handing in both in a test run and in a "normal" run. Make sure you can submit again in case the call doesn't reach the server and that the client loads the existing exam if it reached the server but the response didn't make it to the client. Make sure that submitting after time does not work. (Note that test runs can always be submitted after time.)

Test branches that can be deployed:
- Test a delay in client -> server: https://github.com/ls1intum/Artemis/tree/polish/exam/confirm-end-server-timeout
- Test a delay in server -> client: https://github.com/ls1intum/Artemis/tree/polish/exam/confirm-end-client-timeout

### Screenshots

While (up)loading the submission a spinner indicates something is happening:
![grafik](https://user-images.githubusercontent.com/72132281/107487358-64734e00-6b86-11eb-962b-5b2da0b38e60.png)

An error is displayed for time-outs:
![grafik](https://user-images.githubusercontent.com/72132281/107400482-23d1f100-6b02-11eb-8405-a7acffe6a5ee.png)
